File: kernel/arch/i386/io_port/io_port.cc
#include "kernel/arch/io_port.h"

#include <stdint.h>

namespace arch {

void IoPort::outb(uint8_t output) {
  asm volatile ("outb %0, %1" : /* No outputs */
                              : "a" (output), "Nd" (port_number_)
      );
}

void IoPort::outw(uint16_t output) {
  asm volatile ("outw %0, %1" : /* No outputs */
                              : "a" (output), "Nd" (port_number_)
      );
}

void IoPort::outl(uint32_t output) {
  asm volatile ("outl %0, %1" : /* No outputs */
                              : "a" (output), "Nd" (port_number_)
      );
}

uint8_t IoPort::inb() {
  uint8_t ret;
  asm volatile ("inb %1, %0" : "=r" (ret)
                             : "Nd" (port_number_)
      );
  return ret;
}

uint16_t IoPort::inw() {
  uint16_t ret;
  asm volatile ("inw %1, %0" : "=r" (ret)
                             : "Nd" (port_number_)
      );
  return ret;
}

uint32_t IoPort::inl() {
  uint32_t ret;
  asm volatile ("inl %1, %0" : "=r" (ret)
                             : "Nd" (port_number_)
      );
  return ret;
}

}

File: src/ioport.zig
//! I/O port read/write functions for x86 architecture.

/// Read a byte from the specified I/O port.
pub fn inb(port: u16) u8 {
    var value: u8 = 0;
    asm volatile ("inb %[port], %[value]"
        : [value] "={al}" (value),
        : [port] "{dx}" (port),
        : .{ .memory = true });
    return value;
}

/// Write a byte to an I/O port.
pub fn outb(port: u16, value: u8) void {
    asm volatile ("outb %[value], %[port]"
        :
        : [value] "{al}" (value),
          [port] "{dx}" (port),
        : .{ .memory = true });
}

/// Read a 16-bit value from the specified I/O port.
pub fn inw(port: u16) u16 {
    var value: u8 = 0;
    asm volatile ("inb %[port], %[value]"
        : [value] "={ax}" (value),
        : [port] "{dx}" (port),
        : .{ .memory = true });
    return value;
}

/// Write a 16-bit value to the specified I/O port.
pub fn outw(port: u16, value: u16) void {
    asm volatile ("outw %[value], %[port]"
        :
        : [value] "{ax}" (value),
          [port] "{dx}" (port),
        : .{ .memory = true });
}

/// Read a 32-bit value from the specified I/O port.
pub fn inl(port: u16) u32 {
    var value: u32 = 0;
    asm volatile ("inl %[port], %[value]"
        : [value] "={eax}" (value),
        : [port] "{dx}" (port),
        : .{ .memory = true });
    return value;
}

/// Write a 32-bit value to the specified I/O port.
pub fn outl(port: u16, value: u32) void {
    asm volatile ("outl %[value], %[port]"
        :
        : [value] "{eax}" (value),
          [port] "{dx}" (port),
        : .{ .memory = true });
}
